pep-9999.rst

```diff
@@ -0,0 +1,252 @@
+to the variance of the type parameters of its generic type, 
+``SimpleFunction`` in this example. 
+
+Because variance is restricted to the type parameters of generic types, 
+it makes more sense to specify the variance of type parameters directly
+where the generic type is defined, removing the need to specify it on 
+the type variables themselves. 
+
+
+Proposal
+========
+
+The variance of the type parameters of a generic class can be specified
+with the ``+`` and ``-`` prefix operators on their respective type
+variables within the type parameter list of ``typing.Generic``, 
+``typing.Protocol``, or the generic base class. Instead of
+::
+
+    T_co = typing.TypeVar('T_co', covariant=True)
+    T_contra = typing.TypeVar('T_contra', contravariant=True)
+
+    class SupportsInvertOld(typing.Protocol[T_co]):
+        def __invert__(self) -> T_co: ...
+
+    class SupportsPartialOrderOld(typing.Protocol[T_contra]):
+        def __le__(self, other: T_contra): ...
+
+
+The new syntax uses the ``+`` and ``-`` prefix operators to specify
+variadic type parameters in the same place where the generic type is
+declared:
+::
+
+    T = TypeVar("T")
+    
+    # covariant
+    class SupportsInvert(typing.Protocol[+T]):
+        def __invert__(self) -> T: ...
+
+    # contravariant
+    class SupportsPartialOrder(typing.Protocol[-T]):
+        def __le__(self, other: T): ...
+
+This syntax is inspired by Scala programming language [1]_.
+
+
+
+Specification
+=============
+
+
+Valid use locations
+-------------------
+
+
+The new type variable variance syntax can only be used within the type
+parameter list of ``typing.Generic``, ``typing.Protocol``, or a generic 
+base class. 
+
+The ``+`` and ``-`` prefix operators can be used on ``typing.TypeVar``
+and ``typing.ParamSpec`` instances, that are now already variadic, i.e.
+its ``__covariant__`` and ``__contravariant__`` attributes must be 
+``False``.
+
+When the same type variable is used on multiple generic base classes,
+they must share the same variance, e.g.
+::
+    
+    from typing import TypeVar, Callable, Container, Iterable, Protocol
+    
+    class LinkedList(Iterable[+T], Container[+T]): ...
+    class EventListener(Callable[[-T], None], Protocol[-T]): ...
+    
+are valid examples.
+
+
+Differences with current syntax
+-------------------------------
+
+The new typevar operators return a transparent wrapper around the 
+original type variable, which can be accessed with the ``__origin__``
+attribute on the returned wrapper. e.g.::
+
+    (+T).__origin__ is T
+    (+T).__covariant__ is True
+    (+T).__contravariant__ is False
+    (+T).__name__ == T.__name__
+    (+T).__constraints__ == T.__constraints__
+    (+T).__bound__ is T.__bound__
+
+
+Thus, type variables defined with ``covariant=True`` and 
+``contravariant=True``, are not equivalent to ``+T`` and ``-T``.
+
+
+``+T`` and ``-T`` are not valid type annotations, and should only be
```

Comment:
I suspect type checkers could enforce that on the syntax level without inspecting any of the dunder methods.

Reply:
I guess I'll see when I'll start on the reference implementation for mypy

Reply:
Some type checkers are not written in Python and therefore cannot inspect the runtime state of type vars, e.g. pyre, pyright and PyCharm. These implementation details will be irrelevant for them. mypy of course is another story, but IIUC it also doesn't evaluate type annotations.

Reply:
This section does feel a bit out of place in "Differences with current syntax".